onnxruntime/python/tools/bert/bert_model_optimization.py

```diff
@@ -0,0 +1,778 @@
+        nodes = self.nodes()
+        input_name_to_nodes = self.input_name_to_nodes()
+        output_name_to_nodes = self.output_name_to_nodes()
+        mask_input_name = self.mask_input
+
+        nodes_to_remove = []
+        nodes_to_add = []
+
+        normalize_node = None
+        for node in self.get_normalize_nodes():
+            if self.find_first_parent_by_type(node, self.normalize_name, output_name_to_nodes, recursive=False) is None:
+                normalize_node = node
+                break
+
+        if normalize_node is None:
+            raise Exception("did not find node with op_type", self.normalize_name)
+
+        #This is the first normalize node
+        add_node = self.find_first_parent_by_type(normalize_node, 'Add', output_name_to_nodes, recursive = False)
+        assert(add_node is not None)
+
+        segment_embedding_gather = self.find_first_parent_by_type(normalize_node, 'Gather', output_name_to_nodes, recursive = False)
+        assert(segment_embedding_gather is not None)
+
+        parents = self.get_parents(add_node, output_name_to_nodes)
+        if len(parents) == 2 and parents[0].op_type == 'Gather' and parents[1].op_type == 'Gather':
+            word_embedding_gather = parents[0]
+            position_embedding_gather = parents[1]
+        else:
+            raise Exception("Expect to have two Gather node as parent of Add node")
+
+        input_ids = word_embedding_gather.input[1]
+        segment_ids = segment_embedding_gather.input[1]
+
+        nodes_to_remove, inputs = self.get_parent_nodes_and_inputs(normalize_node, [], output_name_to_nodes)
+
+        embed_node = onnx.helper.make_node('EmbedLayerNormalization',
+                        inputs=[input_ids, segment_ids, mask_input_name, 
+                                word_embedding_gather.input[0], position_embedding_gather.input[0], segment_embedding_gather.input[0],
+                                normalize_node.input[2], normalize_node.input[3]], # gamma and beta
+                        outputs=["embed_output", "mask_idx"],
+                        name="EmbedLayer")
+        embed_node.domain = "com.microsoft"
+        # store embed node for other processing
+        self.embed_node = embed_node
+
+        nodes_to_add.extend([embed_node])
+        nodes_to_remove.append(normalize_node)
+
+        self.replace_input_of_all_nodes(normalize_node.output[0], 'embed_output')
+        self.replace_input_of_all_nodes(mask_input_name, 'mask_idx')
+
+        self.remove_nodes(nodes_to_remove)
+        self.add_nodes(nodes_to_add)
+        self.update_graph(verbose)
+
+    def change_input_to_int32(self):
+        graph = self.graph()
+        inputs = []
+        input_map = {}
+        for input in self.embed_node.input:
+            input_map[input] = onnx.helper.make_tensor_value_info(input, TensorProto.INT32, [self.batch_size if self.batch_size > 0 else 1, self.sequence_length])
+
+        new_graph_inputs = []
```

Comment:
model's input may not be same as embed_node's input.

Reply:
Right. it is fixed in latest commit.